src/safe_get.js

```diff
@@ -7,14 +7,45 @@ function NonExistingPropertyException(object, property) {
   };
 };
 
-module.exports = () => {
-  Object.prototype[Symbol.for('safeGetItem')] = function(property) {
-    if (!(property in this)) {
-      throw new NonExistingPropertyException(this, property);
+const conditionalBind = (func) => {
+  return (object, property) => {
+    const toReturn = func(object, property);
+
+    if (typeof toReturn === 'function') {
+      return toReturn.bind(object);
     } else {
-      return this[property];
+      return toReturn;
     }
-  };
+  }
+};
+
+const safeGetItem = (object, property) => {
+  if (property in object) {
+    return object[property];
+  } else if (Symbol.for('safeGetAttr') in object) {
+    return object[Symbol.for('safeGetAttr')](property);
+  } else if (('has' in object) && (object.has(property))) {
+    return object.get(property);
+  } else {
+    throw new NonExistingPropertyException(object, property);
+  }
+};
 
-  Object.prototype[Symbol.for('safeGetAttr')] = Object[Symbol.for('safeGetItem')];
+const safeGetAttr = (object, property) => {
+  if ('has' in object) {
```

Comment:
// This covers immutablejs, es6 Maps and weak Maps and possible some new custom data structures compliant with this resolution protocol